source/non_matching/mesh_classifier.cc

```diff
@@ -0,0 +1,531 @@
+
+      private:
+        /**
+         * Pointer to the level set function.
+         */
+        const SmartPointer<const Function<dim>> level_set;
+
+        /**
+         * Collection containing the single element which we locally interpolate
+         * the level set function onto.
+         */
+        const hp::FECollection<dim> fe_collection;
+
+        /**
+         * FEValues object used to transform the support points on a cell to
+         * real space.
+         */
+        FEValues<dim> fe_values;
+
+        /**
+         * FEFaceValues object used to transform the support points on a face to
+         * real space.
+         */
+        FEFaceValues<dim> fe_face_values;
+      };
+
+
+
+      template <int dim>
+      AnalyticLevelSetDescription<dim>::AnalyticLevelSetDescription(
+        const Function<dim> &     level_set,
+        const FiniteElement<dim> &element)
+        : level_set(&level_set)
+        , fe_collection(element)
+        , fe_values(element,
+                    Quadrature<dim>(element.get_unit_support_points()),
+                    update_quadrature_points)
+        , fe_face_values(element,
+                         Quadrature<dim - 1>(
+                           element.get_unit_face_support_points()),
+                         update_quadrature_points)
+      {}
+
+
+
+      template <int dim>
+      void
+      AnalyticLevelSetDescription<dim>::get_local_level_set_values(
+        const typename Triangulation<dim>::active_cell_iterator &cell,
+        Vector<double> &local_levelset_values)
+      {
+        fe_values.reinit(cell);
+        const std::vector<Point<dim>> &points =
+          fe_values.get_quadrature_points();
+
+        for (unsigned int i = 0; i < points.size(); i++)
+          local_levelset_values[i] = level_set->value(points[i]);
+      }
+
+
+
+      template <int dim>
+      void
+      AnalyticLevelSetDescription<dim>::get_local_level_set_values(
+        const typename Triangulation<dim>::active_cell_iterator &cell,
+        const unsigned int                                       face_index,
+        Vector<double> &local_levelset_values)
+      {
+        fe_face_values.reinit(cell, face_index);
+        const std::vector<Point<dim>> &points =
+          fe_face_values.get_quadrature_points();
+
+        for (unsigned int i = 0; i < points.size(); i++)
+          local_levelset_values[i] = level_set->value(points[i]);
+      }
+
+
+
+      template <int dim>
+      const hp::FECollection<dim> &
+      AnalyticLevelSetDescription<dim>::get_fe_collection() const
+      {
+        return fe_collection;
+      }
+
+
+
+      template <int dim>
+      unsigned int
+      AnalyticLevelSetDescription<dim>::active_fe_index(
+        const typename Triangulation<dim>::active_cell_iterator &) const
+      {
+        return 0;
+      }
+    } // namespace MeshClassifierImplementation
+  }   // namespace internal
+
+
+
+  using namespace internal::MeshClassifierImplementation;
```

Comment:
Do you think this is worth it? We typically try to avoid too many `using` statements unless there is a very frequent use of the case. But as far as I can see you nicely wrap almost everything into the two classes, any only have the one function `location_from_dof_signs`.